### ISSUE TITLE
More MathFunctions

### ISSUE THREAD

**Issue by nilsvu** [sxs-collaboration/spectre#734]
# Feature request:

For particular systems we would like to specify arbitrary source functions and corresponding analytic solutions in the input file, e.g. for the Poisson equation solver that I am currently building as a first example of an elliptic system. There is currently a `MathFunctions` namespace with a few one-dimensional functions, and a `TensorProduct` to multiply them in higher dimensions. Do we have plans to extend this functionality to support arbitrary multi-dimensional functions in the input file? If so, could we collect a few details here? I'd be willing to implement this then.

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

**Comment by mar-celine** [comment]
As we extend MathFunctions I wonder how much overlap there will be between that and CoordinateMaps, which are essentially multidimensional functions (with the addition of a jacobian and inverses being defined as well). As control system maps are added to CoordinateMaps, maybe it is possible to unify them in some way? Or at least keep things from being redundant/ implemented more than once. There is currently also the AnalyticSolutions directory, is there any particular reason why you want the additional functions to be in MathFunctions as opposed to AnalyticSolutions?

**Comment by nilsvu** [comment]
Yes let's try unify these whenever possible.

The reason why I can't really put this into AnalyticSolutions is that I want to support arbitrary multi-dimensional functions for the source and the solution. AnalyticSolutions are essentially particular (named) compositions of MathFunctions. Instead I'd like to support something like this in the input file:

```yaml
Source:
  Product:
    - Step:
      Onset: -1.0
    - Step:
      Onset: 1.0
      Inverted: True

AnalyticSolution: ...
```

An alternative off the top of my head would be to parse the functions from some format language, e.g. write `step(x-1)*step(-x+1)` instead. Perhaps there are libraries available for this, too.

Whether we want to spend the time on implementing this sort of functionality depends on whether there will be more applications for this. E.g. how will we specify a background metric in the input file for curved-space, fixed-metric executables (would be good to have the possibility to give an arbitrary background)? Are there more use cases where we'd like to support specifying arbitrary sources or constraints and such in the input file?

**Comment by osheamonn** [comment]
We were planning on having more and higher dimensional MathFunctions, it seemed better to wait until we had a proper use case for them. 

I really don't like the idea of having this done by some language on the input files. SpEC allowed math in its input files which led to confusing and error prone input files. The ones you want seem reasonable, but it should definitely be implemented in c++. 

I'm not sure how much the maps and MathFunctions will be similar? Maps are R^n -> R^n and the MathFunctions are R^n -> R. 

Things like the background metric will be separate from the mathfunctions I think, and either be normal input file parameters 

```yaml
BackgroundMetric: 
    KerrSchildBlackHole: 
        Spin: ...
```

or will be specified as part of the executable, and then the individual options to that particular can be specified in the input file? I'm not sure what you mean by arbitrary backgrounds, there's a rather small list of possible backgrounds you can run on isn't there? So you just code up the possible backgrounds you care about and then choose whichever one in the input file.

**Comment by osheamonn** [comment]
For a more constructive comment, what functions were you thinking of? 

* Step function
* Tensor product of functions

What else?

**Comment by nilsvu** [comment]
If all we ever need are a finite set of named functions/solutions/constraints then I'll just work around this by adding a few particular sources and analytic solutions. Only if there will be other use cases at some point where we want to allow for arbitrary functions specified in the input files, then this would be a good opportunity to start working on that. A few specific operations that come to mind (in addition to the obvious `Sum`, `Product`, etc):

- `Spherical` that computes the Euclidean norm of Cartesian coordinates in multiple dimensions and supplies it as the argument to a one-dimensional MathFunction. This would, for instance, allow us to specify an N-dimensional solid sphere _heaviside(1-sqrt(x^2+y^2+z^2))_ as:

  ```yaml
  Spherical:
    Step:
      Onset: 1.0
      Inverted: True
  ```
- `LinearTransformation ` to compute _a*x + b_, the more general `Polynomial` where the arguments are the coefficients to _sum_p(c_p*x^p)_, and the special case `Identity`.

E.g. one solution I'd like to benchmark the elliptic solver with in two dimensions is $u(x, y)=x(1-x)y(1-y)sqrt((x-0.5)^2+(y-0.5)^2)^3$ on a square $x,y \in [0,1]^2$ (this is to compare with Trevor). The operations mentioned above are already sufficient to construct this:

```yaml
Product:
  - TensorProduct:
    Product:
      - Identity:
      - LinearTransformation:
        Scale: -1.0
        Translation: 1.0
  - Pow:
    Exponent: 3.0
    Argument:
      Spherical:
        LinearTransformation:
          Translation: -0.5
```

This syntax is definitely quite verbose, especially for more complex functions. It may still be preferable to a mathematical syntax though, as you mentioned @osheamonn. Are there alternatives?

Instances where functionality like that could be useful (please correct me here, I'd like to get an idea of what functionality we are aiming for) are:

- To provide arguments to a general `SphericallySymmetric` metric, e.g.

  ```yaml
  BackgroundMetric:
    SphericallySymmetric:
      g_rr:
        SomeMathFunctionHere:
      g_tt:
        ...
  ```

- To allow specifying (non-standard) gauge source functions in a generalised harmonic formulation
- Basically wherever we want to allow the user the freedom to input their own functions into precompiled executables, so that they don't have to write code and compile themselves. Since we are writing for a scientific audience this seems useful to me, it helps with exploring beyond what we hard-code into the executables.

**Comment by nilsdeppe** [comment]
I'm going to largely reiterate what @osheamonn said.
- We do NOT want to allow general math in input files. Absolutely not, it was tested in SpEC, it was a good idea that ended up in a horrible mess. Nope, not doing it, absolutely not, at least not while I'm around :) I understand the appeal, but if you really want to do NR, you need to be able to do a bit of `Numerical` too (i.e. C++).
- Why would you ever want to just choose random functions? Like @osheamonn said, there are only a few physically interesting solutions, and they often depend on the PDE structure. Even the spherical example is not interesting in general because you have to do ridiculous compositions where only certain compositions actually satisfy the PDEs.
- Sure, maybe you want to be able to use an arbitrary source, but again, we want physics so no, arbitrary is not what you actually care about :) And again, source terms are specific to the PDE system, so it just makes no sense to support generic things because then they won't be physically relevant.

Unrelated, I guess you could compare to Trevor's code, but don't compare speed, that's just a waste of time.
